### PR TITLE
=clu #18554 Make oldest assignment deterministic when joining

### PR DESCRIPTION
**Pull request by patriknw**
- the reported issue is fixed by the immediate leaderActions
  (moving to Up)  when joining the first node to itself
- the other changes are precautions just in case

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #18554

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala [hunk 1]
```diff
@@ -325,7 +325,7 @@ class ShardRegion(
   def receiveClusterEvent(evt: ClusterDomainEvent): Unit = evt match {
     case MemberUp(m) ⇒
       if (matchingRole(m))
-        changeMembers(membersByAge + m)
+        changeMembers(membersByAge - m + m) // replace
```
Comment (ktoso) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: thanks for the comment, for a second thought was confused ;)

File: akka-cluster-tools/src/test/java/akka/cluster/singleton/ClusterSingletonManagerTest.java [hunk 2]
```diff
@@ -46,77 +46,6 @@ public void demo() {
     //#create-singleton-proxy
   }
 
-  static//documentation of how to keep track of the oldest member in user land
-  //#singleton-proxy
```
Comment (ktoso) [inline akka-cluster-tools/src/test/java/akka/cluster/singleton/ClusterSingletonManagerTest.java]: oh seems it wasn't ever linked to from docs. OK

**Comment by ktoso** [comment]
LGTM.

Would it make sense to mention the `first joining itself will immediately be moved to Up` in the docs at least with a few words?

**Comment by patriknw** [comment]
I think that is an implementation detail, which no happen to coincide with what users always have expected (first node joining is also the oldest and it will of course be moved to Up asap)

**Comment by ktoso** [comment]
True, good as is then :)

File: akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonProxy.scala [hunk 3]
```diff
@@ -203,8 +201,9 @@ final class ClusterSingletonProxy(singletonManagerPath: String, settings: Cluste
    */
   def add(m: Member): Unit = {
     if (matchingRole(m))
-      trackChange {
-        () ⇒ membersByAge += m
+      trackChange { () ⇒
+        membersByAge -= m // replace
```
Comment (viktorklang) [inline akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonProxy.scala]: For consistency?

~~~scala
membersByAge = membersByAge - m + m // replace
~~~

File: akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala [hunk 4]
```diff
@@ -275,7 +275,8 @@ object ClusterEvent {
       val newMembers = newGossip.members -- oldGossip.members
       val membersGroupedByAddress = List(newGossip.members, oldGossip.members).flatten.groupBy(_.uniqueAddress)
       val changedMembers = membersGroupedByAddress collect {
-        case (_, newMember :: oldMember :: Nil) if newMember.status != oldMember.status ⇒ newMember
+        case (_, newMember :: oldMember :: Nil) if newMember.status != oldMember.status || newMember.upNumber != oldMember.upNumber ⇒
+          newMember
       }
       val memberEvents = (newMembers ++ changedMembers) collect {
```
Comment (viktorklang) [inline akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala]: Hmm, didn't that performance article recommend using `diff` and `union` iso -- and ++?

Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala]: @viktorklang yes, that was something I learned from that presentation (surprised me that they are not implemented in the same way). I have changed: https://github.com/akka/akka/commit/811e28595f24de89210ceb85684c93fdb75ce0cb

Comment (viktorklang) [inline akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala]: :+1:

**Pull request merged**